File: src/main/java/findFirstAndLastPositionOfElementInSortedArray/FindFirstAndLastPositionOfElementInSortedArray.java
/*
 * @Source: https://leetcode-cn.com/problems/find-first-and-last-position-of-element-in-sorted-array/
 * @Author: Tian Zhuo
 * @Date: 2019-01-23 21:27:03
 */
package findFirstAndLastPositionOfElementInSortedArray;

public class FindFirstAndLastPositionOfElementInSortedArray {
    public int[] searchRange(int[] nums, int target) {
        if(nums == null || nums.length == 0){
            return new int[]{-1, -1};
        }
        if(nums.length == 1 && nums[0] == target){
            return new int[]{0, 0};
        }
        int[] res = new int[2];
        res[0] = findLeft(nums, 0, nums.length - 1, target);
        res[1] = findRight(nums, 0, nums.length - 1, target);
        return res;
    }

    private int findLeft(int[] nums, int l, int r, int target){
        while(l < r){
            int mid = ((r - l) >> 1) + l;
            if(target > nums[mid]){
                l = mid + 1;
            }else{
                r = mid;
            }
        }
        return nums[l] == target ? l : -1;
    }
    private int findRight(int[] nums, int l, int r, int target){
        while(l < r){
            int mid = ((r - l) >> 1) + l + 1;
            if(target < nums[mid]){
                r = mid - 1;
            }else{
                l = mid;
            }
        }
        return nums[l] == target ? l : -1;
    }
}

File: algorithms/js/findFirstAndLastPositionOfElementInSortedArray/findFirstAndLastPositionOfElementInSortedArray.js
/*
 * @Source: https://leetcode-cn.com/problems/find-first-and-last-position-of-element-in-sorted-array/
 * @Author: Tian Zhuo
 * @Date: 2019-01-23 22:43:28
 */
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var searchRange = function(nums, target) {
    let sl = 0;
    let sm = 0;
    let sr = nums.length - 1;
    let el = 0;
    let em = 0;
    let er = nums.length - 1;
    while (sl < sr || el < er) {
        if (sl < sr) {
            sm = ~~((sr - sl) >> 1) + sl;
            if (nums[sm] < target) {
                sl = sm + 1;
            } else {
                sr = sm;
            }
        }
        
        if (el < er) {
            em = ~~((er - el) >> 1) + el + 1;
            if (nums[em] > target) {
                er = em - 1;
            } else {
                el = em;
            }
        }
    }
    return nums[sl] === target ? [sl, er] : [-1, -1];
};